### DISCUSSION TITLE
Tasmota and a belt winder - How to find the shutter position by roll and reed sensor

### DISCUSSION THREAD

**Discussion by Sebalux** [arendst/Tasmota#14434] [Q&A] [answered]
Hi together,

my recent project is an electrical belt winder which I'm trying to integrate into HA.
Right at the moment I have a Shelly Uni (due to size) running Tasmota.
With that setup I'm able to control the belt winders buttons and move the shutter up and down.
By calibrating the positions and runtime, I think Tasmota and HA will know exactly at what position/percentage the shutter is.

Now to my problem:
The belt winder has two hardware buttons built in for physical control. They need to stay there (waf factor... :D)

So I'm trying to find a solution how Tasmota gets updated and informed when the belt winder moves up/down when one of the physical buttons is pressed.

The beld winder has a belt roll with a magnet and reed switch build in. I assume thats how the belt winder monitors movement and blockage.

I used the ADC pin of the Shelly Uni to bring the reed signal into the Shelly and Tasmota.
GPIO 17 of the Shelly Uni is configured as ADC Range in Tasmota. That brings me 0 or a value of about 330 every time the magnet triggers the reed switch.

So how can I use that information for Tasmota to detect that the shutter is moving, if its moving up/down (impossible I think) and how far it moved? I'm afraid that's just impossible without bigger changes. Maybe someone has a clever idea for me?

Thanks!

**Comment by rakaiser** [comment, marked as answer]
Hi  Sebalux,

I started similar project pimping my 4 x GW60 belt winders with opto couplers for up/down button and managed to put a reed contact in cavity near the belt roll. I made own 2 layer PCB with ESP12, 3.3V step down to use 24V supply of winder and several sensor module options (PIR, IR, RF-RX, several IIC) to fit in some small standard housing (should have thought  more about mechanical tolerances ).

I have the reed switching to GND at normal GPIO with pull up to 3.3V and configured as counter. That work properly giving some 50 ticks per full run. Analog I use for illuminance. But as You stated there is no info about direction with only single reed (was also my fallacy).  
The counter could potentially be used for positioning instead of time based calibration. But this works very well and code ist quite complex. 
So without wiring the wiring the buttons from the winder to Tasmota device the only option I see is to use counter to check for manual interaction from HA side. E. g. set counter to 0 after any action triggered by HA has finished. If its <>0 at next HA action then it has been moved manually and should set first to end stop nearest requested target. 

BTW-1, there is/was a bug in current implementation for some time with the required mode 1 with pulse relays mode. Problem is/was that for stop always 2nd relay is triggered leading shutter to go down when upper end stop has been reached before.
This is fixed now in development by pull request 14405 / @stefanbode.  

BTW-2, I realized that in my use case the GPIO used are relevant. With GPIO12/14 my shutter goes always up at power on due high level at initial boot. I guess only GPIO4/5 are low.

**Reply by stefanbode** [reply]
Always interesting what people finally do with the driver. Currently only the stepper mode uses the counter to measure real position. On the relay level you get one relay for on/off and another one for direction. I assume for a belt drive this is only partially helpful. I would take a look at the failsafe configuration with two switching relays. Currently I would assume that this can be wired in way that one relay do up and one down. The pwm signal in this case you do not need

**Reply by rakaiser** [reply]
Would have used something else two years ago when I renovated roller shutters / belts. But found nothing for HA to fit in standard wall cut off. Meanwhile there is at least one from Rollotron for DECT / Fritzbox.
Concerning your proposal to look at failsafe mode I didn't get what you mean? IMHO only with mode 1 SHT_TIME creates required start stop pulses on correct outputs. Am I wrong?

**Reply by stefanbode** [reply]
I was thinking about your counter, to use this for operation. The time_based shutter is always fine this also works with the pulse mode. Now I also understand that you HAVE TO use the pulse mode. This are not normal motors where you just power on and as long as the relay is ON the shutter operates. For the normal motors it would be possible to "mis-use" the the stepper motor functionality. But this for sure does not work if you need the pulses to start and stop.

**Reply by stefanbode** [reply]
But maybe there is a more general approach required that you can add a counter to all modes and enhance position quality. I take a look at the Rollotron. From my perspective this should be fine with time based operation anyhow

**Reply by stefanbode** [reply]
Regarding your high after reboot please check the End of this page which gpio are high or low by default. Sometimes a 10k resistor can help. https://tasmota.github.io/docs/DeepSleep/#esp8266-deepsleep-side-effects

**Reply by stefanbode** [reply]
I checked the code for options to include a counter for defining the position even if in time mode though a pulse counter. In time mode there are 20 steps per second. Ideally the counter should do 10..50 to avoid massive rounding effects. Therefore if getting up to 50 pulses per second I can make a small change and we can give it a try. It would work in the way that when you define shutteropenduration 12 this are 12x20=240 steps. The shutter will then operate until the counter show 240. closeduration must be the same = 12.

Interested and able to test?

**Reply by rakaiser** [reply]
Concerning initial boot up I stick to this: https://rabbithole.wwwdotorg.org/2017/03/28/esp8266-gpio.html
Not sure if cold boot and deep sleep recovery is same but GPIO 4/5 seems the only to be really safe. Looks my PCBs need some more handwiring fixes.

The counter is driven by 2 magnets in the belt roll with some 2cm diameter giving 54 ticks per run only. So the counter is linear to the belt movement which is not fully proportional to the shutter position. This is because diameter of the upper belt sleeve is changing. I realized that when having calibrated against the counter values at 30, 50, 70, 90, 100%. At 50% opening it is now rather one third only. 
So 4 me time calibration is rather better and accuracy is good enough.

**Reply by stefanbode** [reply]
Calibration will still work also with the counter. I just will transform the counter into the time domain to make as less changes as possible and do not break anything. Anyhow 54 counts for the complete movement is very low. I'm not able to predict if this could work either. Regarding the initial question of the hardware buttons and how to get this info into the driver. The only way is to wire these to two GPIO and the use as normal buttons. And there will be an additional problem because in this configuration the hardware switches the relay either. Not very good. I would disconnect the buttons from the original device and just let Tasmota do the job.

**Reply by rakaiser** [reply]
I could test a solution with the counter but as stated for me the existing solution is pretty fine. 
I decided not to disconnect the switches because it would have broken the autonomous function of the device without Tasmota. Maybe it's an option for @Sebalux. I'll check counter to detect manual move.
Also had a look in to the manual of Rollotron DECT device. There is no user defined calibration at all and only a favourite position is mentioned. A belt winder with an ESP and Tasmota would be an interesting device ...

**Reply by stefanbode** [reply]
Ok in this case i leave it as it is. Maybe later this month I have some time to test this on my own and check for best precise and implementation.
